File: src/api-service/__app__/onefuzzlib/updates.py
#!/usr/bin/env python
#
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

import logging
from typing import Dict, Optional, Type

from azure.core.exceptions import ResourceNotFoundError
from msrestazure.azure_exceptions import CloudError
from onefuzztypes.enums import UpdateType
from pydantic import BaseModel

from .azure.queue import queue_object
from .azure.storage import StorageType


# This class isn't intended to be shared outside of the service
class Update(BaseModel):
    update_type: UpdateType
    PartitionKey: Optional[str]
    RowKey: Optional[str]
    method: Optional[str]


def queue_update(
    update_type: UpdateType,
    PartitionKey: Optional[str] = None,
    RowKey: Optional[str] = None,
    method: Optional[str] = None,
    visibility_timeout: int = None,
) -> None:
    logging.info(
        "queuing type:%s id:[%s,%s] method:%s timeout: %s",
        update_type.name,
        PartitionKey,
        RowKey,
        method,
        visibility_timeout,
    )

    update = Update(
        update_type=update_type, PartitionKey=PartitionKey, RowKey=RowKey, method=method
    )

    try:
        if not queue_object(
            "update-queue",
            update,
            StorageType.config,
            visibility_timeout=visibility_timeout,
        ):
            logging.error("unable to queue update")
    except (CloudError, ResourceNotFoundError) as err:
        logging.error("GOT ERROR %s", repr(err))
        logging.error("GOT ERROR %s", dir(err))
        raise err


def execute_update(update: Update) -> None:
    from .jobs import Job
    from .orm import ORMMixin
    from .pools import Node, Pool, Scaleset
    from .proxy import Proxy
    from .repro import Repro
    from .tasks.main import Task

    update_objects: Dict[UpdateType, Type[ORMMixin]] = {
        UpdateType.Task: Task,
        UpdateType.Job: Job,
        UpdateType.Repro: Repro,
        UpdateType.Proxy: Proxy,
        UpdateType.Pool: Pool,
        UpdateType.Node: Node,
        UpdateType.Scaleset: Scaleset,
    }

    # TODO: remove these from being queued, these updates are handled elsewhere
    if update.update_type == UpdateType.Scaleset:
        return

    if update.update_type in update_objects:
        if update.PartitionKey is None or update.RowKey is None:
            raise Exception("unsupported update: %s" % update)

        obj = update_objects[update.update_type].get(update.PartitionKey, update.RowKey)
        if not obj:
            logging.error("unable find to obj to update %s", update)
            return

        if update.method and hasattr(obj, update.method):
            logging.info("performing queued update: %s", update)
            getattr(obj, update.method)()
            return
        else:
            state = getattr(obj, "state", None)
            if state is None:
                logging.error("queued update for object without state: %s", update)
                return
            func = getattr(obj, state.name, None)
            if func is None:
                logging.debug(
                    "no function to implement state: %s - %s", update, state.name
                )
                return
            logging.info(
                "performing queued update for state: %s - %s", update, state.name
            )
            func()
        return

    raise NotImplementedError("unimplemented update type: %s" % update.update_type.name)

File: src/ApiService/ApiService/onefuzzlib/Updates.cs
using Microsoft.OneFuzz.Service.OneFuzzLib.Orm;
using ApiService.OneFuzzLib.Orm;
using System;
using System.Threading.Tasks;
using System.Linq;

namespace Microsoft.OneFuzz.Service;

public interface IUpdates
{
    public Task ExecuteUpdate(Update update);
}

public class Updates : IUpdates
{
    private readonly ILogTracer _logger;
    IProxyOperations _proxyOperations;
    INodeOperations _nodeOperations;
    public Updates(ILogTracerFactory loggerFactory, IProxyOperations proxyOperations, INodeOperations nodeOperations)
    {
        _logger = loggerFactory.MakeLogTracer(Guid.NewGuid());
        _proxyOperations = proxyOperations;
        _nodeOperations = nodeOperations;
    }

    public async Task ExecuteUpdate(Update update)
    {
        switch (update.UpdateType)
        {
            // TODO: Task, Job, Repro, Pool
            case UpdateType.Proxy:
                await GenericUpdate<Proxy, IProxyOperations>(update, _proxyOperations);
                break;
            case UpdateType.Node:
                await GenericUpdate<Node, INodeOperations>(update, _nodeOperations);
                break;
            case UpdateType.Scaleset:
                return;
            default:
                throw new NotImplementedException($"unimplemented update type: {update.UpdateType}");
        }
    }

    private async Task GenericUpdate<TEntity, TOperation>(Update update, TOperation operation)
        where TEntity : EntityBase
        where TOperation : IOrm<TEntity>
    {
        if (update.ParititionKey == null || update.RowKey == null)
        {
            throw new ArgumentException($"unsupported update {update}");
        }

        var obj = await operation.QueryAsync(filter: $"RowKey eq '{update.RowKey}' and PartitionKey eq '{update.ParititionKey}'").FirstOrDefaultAsync();
        if (obj == null)
        {
            _logger.Error($"unable to find obj to update {update}");
            return;
        }

        if (update.Method != null)
        {
            var method = obj.GetType().GetMethod(update.Method);
            if (method != null)
            {
                _logger.Info($"performing queued update: {update}");
                method.Invoke(null, null);
                return;
            }
        }
        else
        {
            var state = obj.GetType().GetMember("state");
            if (state == null)
            {
                _logger.Error($"queued update for object without state: {update}");
                return;
            }

            var func = state.GetType().GetMethod("name");
            if (func == null)
            {
                // TODO: We need debug in logger
                _logger.Info($"no function to implement state: {update} - {state}");
                return;
            }
            _logger.Info($"performing queued update for state: {update} - {func.Name}");

            func.Invoke(null, null);
        }
        return;
    }
}
